airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java

```diff
@@ -0,0 +1,189 @@
+    if (!areBothEmpty(expectedObject, actualObject)) {
+      LOGGER.info("Expected Object : {}", expectedObject);
+      LOGGER.info("Actual Object   : {}", actualObject);
+      final Iterator<Map.Entry<String, JsonNode>> expectedDataIterator = expectedObject.fields();
+      assertEquals(expectedObject.size(), actualObject.size(), "Unequal row size");
+      while (expectedDataIterator.hasNext()) {
+        final Map.Entry<String, JsonNode> expectedEntry = expectedDataIterator.next();
+        final JsonNode expectedValue = expectedEntry.getValue();
+        String key = expectedEntry.getKey();
+        JsonNode actualValue = ComparatorUtils.getActualValueByExpectedKey(key, actualObject, this::resolveIdentifier);
+        LOGGER.info("For {} Expected {} vs Actual {}", key, expectedValue, actualValue);
+        assertSameValue(expectedValue, actualValue);
+      }
+    } else {
+      LOGGER.info("Both rows are empty.");
+    }
+  }
+
+  private boolean areBothEmpty(final JsonNode expectedData, final JsonNode actualData) {
+    if (expectedData.size() <= 1 && actualData.size() <= 1) {
+      boolean isExpectedEmpty = (expectedData.size() == 0 || expectedData.iterator().next().asText().isEmpty());
+      return isExpectedEmpty && (actualData.size() == 0 || actualData.iterator().next().asText().isEmpty());
+    } else {
+      return false;
+    }
+  }
+
+  // Allows subclasses to implement custom comparison asserts
+  protected void assertSameValue(final JsonNode expectedValue, final JsonNode actualValue) {
+    LOGGER.info("assertSameValue : {} vs {}", expectedValue, actualValue);
+
+    assertTrue(compareJsonNodes(expectedValue, actualValue), "Expected value " + expectedValue + " vs Actual value " + actualValue);
+  }
+
+  protected boolean compareJsonNodes(final JsonNode expectedValue, final JsonNode actualValue) {
+    Boolean t = true;
+    if (expectedValue == null || actualValue == null) {
+      return expectedValue == null && actualValue == null;
+    } else if (expectedValue.isNumber() || expectedValue.isDouble() || expectedValue.isFloat()) {
+      return compareNumericValues(expectedValue.asText(), actualValue.asText());
+    } else if (expectedValue.isBoolean()) {
+      return compareBooleanValues(expectedValue.asText(), actualValue.asText());
+    } else if (isDateTimeWithTzValue(expectedValue.asText())) {
+      return compareDateTimeWithTzValues(expectedValue.asText(), actualValue.asText());
+    } else if (isDateTimeValue(expectedValue.asText())) {
+      return compareDateTimeValues(expectedValue.asText(), actualValue.asText());
+    } else if (isDateValue(expectedValue.asText())) {
+      return compareDateValues(expectedValue.asText(), actualValue.asText());
+    } else if (expectedValue.isArray() && actualValue.isArray()) {
+      return compareArrays(expectedValue, actualValue);
+    } else if (expectedValue.isObject() && actualValue.isObject()) {
+      compareObjects(expectedValue, actualValue);
+      return true;
```

Comment:
is it expected to return `true` every time?

Reply:
the `compareObjects` contains asserts inside.
I could return the result as boolean but I decided that it's better to raise exceptions on the level where the issue is. Otherwise, we will get an error that something is wrong on the top level.